src/silly_sprites/opengl/shader.cpp

```diff
@@ -1,65 +1,76 @@
+        }
+        auto const trimmed_source = trim(std::string{ source });
+        if (trimmed_source.empty()) {
+            spdlog::warn("empty shader source");
+            return tl::unexpected{ GlError{ GlErrorType::EmptyShaderSource } };
+        }
         auto const error_message = [&](GLuint id) -> std::string {
             auto len = GLint{};
             glGetShaderiv(
                     id,
                     GL_INFO_LOG_LENGTH,
                     &len
             );      // gets the length of the error message including the null terminator
-                    // cast beacuse the gl function needs the signed int and the string ctor expects a unsigned int.
-                    // this causes a waring on linux because the sign could chance.
-            auto message = std::string(static_cast<usize>(len - 1), ' '); // create a string with a suitable length
+                    // cast because the gl function needs the signed int and the string ctor expects a unsigned int.
+                    // this causes a warning on linux because the sign could chance.
+            auto message = std::string(gsl::narrow_cast<usize>(len - 1), ' '); // create a string with a suitable length
             glGetShaderInfoLog(id, len - 1, nullptr, message.data());
-            return trim(message);
+            return trim(std::move(message));
         };
 
-        m_name = glCreateShader(sly::to_underlying(type));
-        char const* c_source = source.data();
-        glShaderSource(m_name, 1, &c_source, nullptr);
-        glCompileShader(m_name);
+        auto const name = glCreateShader(sly::to_underlying(type));
+        if (name == 0) {
+            spdlog::critical("Failed to create Shader");
+            return tl::unexpected{ GlError{ GlErrorType::FailedToCreateShader } };
+        }
+        auto const sources = std::array{ source.data() };
+        glShaderSource(name, gsl::narrow_cast<GLsizei>(sources.size()), sources.data(), nullptr);
+        glCompileShader(name);
 
         auto result = GLint{};
-        glGetShaderiv(m_name, GL_COMPILE_STATUS, &result);
+        glGetShaderiv(name, GL_COMPILE_STATUS, &result);
 
-        if (result == GL_TRUE) {
-            m_valid = true;
-            spdlog::info("SUCCESS::SHADER::{}::COMPILATION", get_name_from_shader_type(type));
-        } else {
-            m_valid = false;
-            auto message = error_message(m_name);
+        if (result != GL_TRUE) {
+            auto const message = error_message(name);
             spdlog::critical(
                     "ERROR::SHADER::{}::COMPILATION_FAILED\nerror: {}",
                     get_name_from_shader_type(type),
                     message
             );
+            return tl::unexpected{ GlError{ GlErrorType::FailedToCompileShader } };
         }
+        spdlog::info("SUCCESS::SHADER::{}::COMPILATION", get_name_from_shader_type(type));
+        return { Shader{ name } };
```

Comment:
I think the outer braces are unnecessary.
```suggestion
        return Shader{ name };
```